### PR TITLE
TS-4321: Keep response headers in FetchSM as they are

### PR DESCRIPTION
**Pull request by maskit**
https://issues.apache.org/jira/browse/TS-4321

### CODE REVIEW

**Comment by jpeach** [comment]
Is this testable? It seems like we really need a regression test here ...

**Comment by maskit** [comment]
Though I tested it with nghttp and httpbin as an origin server, I think we can test with nghttp/curl and nc. But if you mean automated tests, I have no idea.

What do you think adding request/response data files as a first step so that we can test manually with nc and/or curl?

**Comment by maskit** [comment]
I just remembered tsqa. I'll try to write a test for this with tsqa.

File: proxy/http2/HTTP2.cc [hunk 1]
```diff
@@ -498,25 +498,23 @@ http2_convert_header_from_2_to_1_1(HTTPHdr *headers)
   return PARSE_DONE;
 }
 
-void
-http2_convert_header_from_1_1_to_2(HTTPHdr *headers)
+HTTPHdr *
+http2_generate_h2_header_from_1_1(HTTPHdr *headers)
```
Comment (jpeach) [inline proxy/http2/HTTP2.cc]: Make ``headers`` since you are not changing it here.

File: proxy/http2/Http2ConnectionState.cc [hunk 2]
```diff
@@ -1007,14 +1007,18 @@ Http2ConnectionState::send_headers_frame(FetchSM *fetch_sm)
 
   DebugHttp2Stream(ua_session, stream->get_id(), "Send HEADERS frame");
 
-  http2_convert_header_from_1_1_to_2(resp_header);
-  buf_len = resp_header->length_get() * 2; // Make it double just in case
+  HTTPHdr *h2_hdr = http2_generate_h2_header_from_1_1(resp_header);
```
Comment (jpeach) [inline proxy/http2/Http2ConnectionState.cc]: I think we should probably avoid the malloc by using an out parameter for the HTTP2 copy.

**Comment by zwoop** [comment]
Assigned to peach for review, and we really need this landed for 6.2.0, because it's completely crippling.

File: proxy/http2/HTTP2.cc (hunk 1 above)
Comment (maskit) [inline proxy/http2/HTTP2.cc]: I wanted to do so. However, ``iter_get_first`` doesn't have ``const``.

**Comment by maskit** [comment]
@jpeach I added a parameter to ``http2_generate_h2_header_from_1_1()`` to pass a ``HTTPHdr`` declared in ``send_headers_frame()`` instead of creating it with ``new`` in ``http2_generate_h2_header_from_1_1()``. Do I understand correctly?

I tried to write a test with tsqa but I haven't understood the test framework yet and some tests fail on my box. The test we need here is just send a request with HTTP/2, return a chunked response from an origin server, and then check the response body size. Could you help me out?

I think we should probably land this fix first without the regression test, because it's a 6.2.0 blocker.

**Comment by zwoop** [comment]
I don't think we should hold this back to enforce a tsqa test. However, I'd suggest you file a separate Jira to add such a test. Curious: Does any of the other H2 test suites detect this? And is there a reasonable way to hook up those test suites to tsqa (i.e. a wrapper over those existing tests?).

**Comment by zwoop** [comment]
Fwiw, I'm running with this patch now on {ci,docs}.trafficserver.apache.org, and it's fixing the issues I've seen on e.g. https://ci.trafficserver.apache.org/

**Comment by maskit** [comment]
I'm going to push this fix without the test tomorrow, and I'll file a Jira for tests as a more general (but not too general) issue. Since tsqa has a test for chunked responses, we should reuse it.

I don't know test suites which can detect this. It seems that tsqa wraps h2spec, though it doesn't detect this.

Also, I think FetchSM should expose the headers with ``const`` modifier. We would need to make HTTPHdr usable with ``const`` first. I'll file a JIra for these too.

**Comment by jpeach** [comment]
@maskit sound good to me

**Comment by maskit** [comment]
I filed three Jiras.
https://issues.apache.org/jira/browse/TS-4349
https://issues.apache.org/jira/browse/TS-4350
https://issues.apache.org/jira/browse/TS-4351

**Pull request merged**